pkg/applicationserver/io/pubsub/pubsub.go

```diff
@@ -18,118 +18,279 @@ package pubsub
 import (
 	"context"
 	"fmt"
+	"sync"
+	"time"
 
 	"go.thethings.network/lorawan-stack/pkg/applicationserver/io"
-	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/formatters"
+	"go.thethings.network/lorawan-stack/pkg/applicationserver/io/pubsub/provider"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/errorcontext"
+	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/log"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/unique"
 	"gocloud.dev/pubsub"
 )
 
-type srv struct {
-	ctx           context.Context
-	server        io.Server
-	formatter     formatters.Formatter
-	topics        []*pubsub.Topic
-	subscriptions []*pubsub.Subscription
+// PubSub is an PubSub frontend that exposes ttnpb.ApplicationPubSubRegistryServer.
+type PubSub struct {
+	ttnpb.ApplicationPubSubRegistryServer
+
+	*component.Component
+	ctx      context.Context
+	server   io.Server
+	registry Registry
+
+	integrations      sync.Map
+	integrationErrors sync.Map
 }
 
-// Start starts the pusub frontend.
-func Start(ctx context.Context, server io.Server, formatter formatters.Formatter, pubURLs, subURLs []string) ([]*io.Subscription, error) {
-	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/pubsub")
-	s := &srv{
+// New creates a new pusub frontend.
+func New(c *component.Component, server io.Server, registry Registry) (*PubSub, error) {
+	ctx := log.NewContextWithField(c.FillContext(c.Context()), "namespace", "applicationserver/io/pubsub")
+	ps := &PubSub{
+		Component: c,
 		ctx:       ctx,
 		server:    server,
-		formatter: formatter,
+		registry:  registry,
 	}
+	ps.RegisterTask(ctx, "integrate_all", ps.integrateAll, component.TaskRestartOnFailure)
+	return ps, nil
+}
+
+func (ps *PubSub) integrateAll(ctx context.Context) error {
+	return ps.registry.Range(ctx, []string{"ids"},
+		func(ctx context.Context, _ ttnpb.ApplicationIdentifiers, pb *ttnpb.ApplicationPubSub) bool {
+			ps.startIntegrationTask(ctx, pb.ApplicationPubSubIdentifiers)
+			return true
+		},
+	)
+}
+
+var integrationBackoff = []time.Duration{100 * time.Millisecond, 1 * time.Second, 10 * time.Second}
 
-	subs := []*io.Subscription{}
-	for _, pubURL := range pubURLs {
-		topic, err := pubsub.OpenTopic(ctx, pubURL)
+func (ps *PubSub) startIntegrationTask(ctx context.Context, ids ttnpb.ApplicationPubSubIdentifiers) {
+	ctx = log.NewContextWithFields(ctx, log.Fields(
+		"application_uid", unique.ID(ctx, ids.ApplicationIdentifiers),
+		"pub_sub_id", ids.PubSubID,
+	))
+	ps.StartTask(ctx, "integrate", func(ctx context.Context) error {
```

Comment:
`pubsub`; we probably need the UID here, right? Don't we do that in AS linking too?

Reply:
Not needed - you can use the `ApplicationPubSubIdentifiers` to get the actual PubSub inside the task.

Reply:
AS linking needs application UIDs since the links are indexed on a per-application basis.

Reply:
I meant the task ID, but indeed this is not necessary